base/channels.jl

```diff
@@ -120,12 +124,16 @@ julia> String(collect(chnl))
 "hello world"
 ```
 """
-function Channel{T}(func::Function, size=0; taskref=nothing) where T
+function Channel{T}(func::Function, size=0; taskref=nothing, spawn=false) where T
```

Comment:
Unfortunately the deprecated `Channel(func::Function; ctype, csize, taskref)` definition overwrites the `Channel(::Function)` case, so `spawn` can't be used with it. I guess it's simplest to add `spawn` to that definition.

Reply:
Ah shoot, you're right. :'(

I had previously added `spawn=` to the deprecated constructor when i originally wrote this a few weeks ago, but when i was cleaning it up now I was confused by that and removed it. I thought this would be a good carrot to bring people over to the new constructor syntax in order to use the `spawn=true` argument...


Hmm, is it a crazy/bad idea to add the `spawn` argument to the deprecated constructor and then also add a runtime check that the user doesn't set `spawn=true` _and_ any of the old kwargs, and if they do, throw an ArgumentError? (or fake a MethodError)?

Something like:
```julia
function Channel(func::Function; ctype=nothing, csize=nothing, taskref=nothing, spawn=nothing)
    if (ctype !== nothing || csize !== nothing) && spawn !== nothing
        throw(ArgumentError("Cannot set `spawn=` in the deprecated constructor `Channel(f; ctype=Any, csize=0)`. Please use `Channel{Any}(f, size=0; taskref=nothing, spawn=false)` instead!"))
    end
    # Set default values
    ctype === nothing && ctype=Any
    csize === nothing && csize=0
    spawn === nothing && spawn=false
    return Channel{ctype}(func, csize; taskref=taskref)
end
```

This would allow us to bring the users along to the new constructor with a "soft deprecation" of the old method: "if you want to use the new feature, you need to use the new method", in exchange for a small runtime cost. (And I think I've arranged the short-circuiting correctly above so that the correct case `Channel(spawn=true) do ch; ... end` is the fastest.)

What do you think? Is that overkill?

Reply:
Okay i'm doing this.